`run_if` for `SystemConfigs` via anonymous system sets

geieredgar opened the pull request:
# Objective

- Fixes #7659

## Solution

The idea of anonymous system sets or "implicit hidden organizational sets" was briefly mentioned by @cart here: https://github.com/bevyengine/bevy/pull/7634#issuecomment-1428619449.

- `Schedule::add_systems` creates an implicit, anonymous system set of all systems in `SystemConfigs`.  
-  All dependencies and conditions from the `SystemConfigs` are now applied to the implicit system set, instead of being applied to each individual system. This should not change the behavior, AFAIU, because `before`, `after`, `run_if` and `ambiguous_with` are transitive properties. 
- The newly added `AnonymousSystemSet` stores the names of its members to provide better error messages.
- The names are stored in a reference counted slice, allowing fast clones of the `AnonymousSystemSet`.
- However, only the pointer of the slice is used for hash and equality operations
  - This ensures that two `AnonymousSystemSet` are not equal, even if they have the same members / member names.
     - So two identical `add_systems` calls will produce two different `AnonymousSystemSet`s.
  - Clones of the same `AnonymousSystemSet` will be equal.

## Drawbacks
If my assumptions are correct, the observed behavior should stay the same. But the number of system sets in the `Schedule` will increase with each `add_systems` call. If this has negative performance implications, `add_systems` could be changed to only create the implicit system set if necessary / when a run condition was added.

alice-i-cecile commented:
Oh very clever. I'll do a full review and add this to the milestone. Welcome and thank you!

hymm commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -371,8 +371,17 @@ impl ScheduleGraph {
     fn add_systems<P>(&mut self, systems: impl IntoSystemConfigs<P>) {
-        let SystemConfigs { systems, chained } = systems.into_configs();
-        let mut system_iter = systems.into_iter();
+        let SystemConfigs {
```
what's the difference between `add_systems` and `configure_set` with this change?

geieredgar replied:
I am not exactly sure what you mean, because I would say the difference between these two did not change.
`add_systems` operates on an implicit system set defined by the added systems and `configure_set` on an explicitly passed system set, but without adding any systems.

hymm replied:
if run_systems is just using configure_set under the hood now is there possibly a way to combine the apis?

geieredgar replied:
Hm, I don't think so. You can't combine this approach with a system set that is accessible/passed from outside, because it would have unexpected side effects. E.g. the current  `add_systems((a,b).after(c).in_set(d))` would not be equal to an adjusted `add_systems((a, b).after(c), d)` when we use `d` instead of the current anonymous set, because `d` after `c` is only implied by the latter. But maybe I am still misunderstanding something.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -112,6 +114,58 @@ impl<T> SystemSet for SystemTypeSet<T> {
+/// The names of the members are stored in a reference counted slice.
+/// The pointer to the slice is used as hash value and for the equality checks.
+pub struct AnonymousSystemSet {
+    system_names: Arc<[Cow<'static, str>]>,
```
I'd like a pub accessor for this field for debug / viz purposes.

geieredgar replied:
Done.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -112,6 +114,58 @@ impl<T> SystemSet for SystemTypeSet<T> {
+
+impl Hash for AnonymousSystemSet {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::as_ptr(&self.system_names).hash(state);
```
This needs a comment explaining the details you captured in the PR description. This is critical logic, and easy to overlook.

geieredgar replied:
Okay, I have added comments to both the Hash and PartialEq implementation and moved the PartialEq implementation closer to the Hash implementation to make it harder to overlook.

alice-i-cecile reviewed (changes requested):
Generally very impressed. I'd like to see some basic tests for hashing and equality added, based on the behavior in the PR description.

cart commented:
Haven't done a full review yet, but I do think this is _probably_ the right path forward. But I'm very curious what others think (especially @maniwani, given their work in this area).

First some pros:
1. It makes run conditions "more optimal" by only running them once for a group of systems. (also a con in some contexts).
2. It makes it possible to express `(foo, bar).run_if(X)`, which might  help make removing OnUpdate possible / desirable.
3. It makes our graphs smaller and therefore faster to "crunch". If we have 100 systems all sharing a `after(X)` constraint,  thats 99 fewer edges in the graph.

Some cons:
1. This is very implicit. There is a behavioral difference between `.add_system(foo.run_if(X)).add_system(bar.run_if(X))` vs `.add_systems((foo, bar).run_if(X))`, which will be more pronounced the more "separate" `foo` and `bar` are in a schedule.
2. This makes "schedule visualization" more difficult as rendering these set names could be very unwieldy for large numbers of systems in a set.
3. As already mentioned, this has the potential to create a lot of "useless" sets. I think we should avoid creating these sets when they aren't needed (ex: they dont have before/after/run_if/in_set).
4. Idk if including the system names in the anonymous sets is ideal. We can already do "set membership" calculations elsewhere if we need this info, and any "set debug" view will already show the members of systems in a set. Additionally, it is a `O(N)` allocation that is only useful in a debug context. Probably better to just give anonymous sets ascending ids or something. Ex: `AnonymousSet(5)`. We should double check that this won't negatively affect any of our error messages though.

alice-i-cecile commented:
> This is very implicit. There is a behavioral difference between .add_system(foo.run_if(X)).add_system(bar.run_if(X)) vs .add_systems((foo, bar).run_if(X)), which will be more pronounced the more "separate" foo and bar are in a schedule.

I'd like to consolidate this in the future, and make *all* instances of run conditions with the same name get checked exactly once. Users who want different timing can newtype. I think that the way that each `run_if` is evaluated seperately in *some* circumstances is already a footgun in confusing ways.

> This makes "schedule visualization" more difficult as rendering these set names could be very unwieldy for large numbers of systems in a set.

Agreed, I'm feeling a bit iffy about the naming questions here.

> As already mentioned, this has the potential to create a lot of "useless" sets. I think we should avoid creating these sets when they aren't needed (ex: they dont have before/after/run_if/in_set).

Agreed. Now that I'm thinking more about how this works, I think this should be fixed here to avoid making a complete mess of scheduling viz.

geieredgar commented:
> >  As already mentioned, this has the potential to create a lot of "useless" sets. I think we should avoid creating these sets when they aren't needed (ex: they dont have before/after/run_if/in_set).

> Agreed. Now that I'm thinking more about how this works, I think this should be fixed here to avoid making a complete mess of scheduling viz.

Okay, I adjusted `add_systems` to only create an anonymous system set if at least one run condition was added. This should limit the impact of this PR on scheduling viz as long as `run_if` is not used.  

>  >  This makes "schedule visualization" more difficult as rendering these set names could be very unwieldy for large numbers of systems in a set.

> Agreed, I'm feeling a bit iffy about the naming questions here.

I agree that for schedule visualizations detailed names aren't necessary because they are redundant if you can already see the members in the visualization.

> 4. Idk if including the system names in the anonymous sets is ideal. We can already do "set membership" calculations elsewhere if we need this info, and any "set debug" view will already show the members of systems in a set. Additionally, it is a `O(N)` allocation that is only useful in a debug context. Probably better to just give anonymous sets ascending ids or something. Ex: `AnonymousSet(5)`. We should double check that this won't negatively affect any of our error messages though.

I have switched to using an incremented atomic counter for the implicit sets. I already have thought about that, but was worried about the error messages. But now I know that we can simply reconstruct the detailed set name when an  error occurs. But I would like to do this in an separate PR.

What do you think about adding an `is_anonymous()` method to the `SystemSet` trait? I want to do this to distinguish between using the Debug trait for the name and constructing the name from the set members for the error message code. But this might also be useful for the visualization to skip/unpack those sets.

geieredgar commented:
Summarization of my latest changes:

- I've made sure that error messages for a call `add_systems((a, b))` would print out the anonymous set as `"(<name of a>, <name of b>)"` instead of `"AnonymousSet(<ID>)"`. To do that, I temporarily removed the `Debug` trait from `SystemSet` and looked where is was used. 
- Added `is_anonymous()` method to `SystemSet` to be able to distinguish between anonymous and non-anonymous system sets. The schedule visualization should use this information to directly apply all dependencies and relations from the system set to its members, to make the anonymous set invisible in the visualization. 
- Renamed `AnonymousSystemSet` to `AnonymousSet`, I like the shorter name better.
- Anonymous sets are now always created except in the two simple cases of `add_systems((a, b, ...))` and `add_systems((a, b, ...).chain())`. This made the implementation simpler, because we don't have to copy from `graph_info` of the whole`SystemConfigs` into each `SystemConfig`. It also retains the benefits of a smaller schedule graph mentioned by cart.

With this, I think every mentioned concern excluding the behavioral difference between `.add_system(foo.run_if(X)).add_system(bar.run_if(X))` and `.add_systems((foo, bar).run_if(X))` is addressed.

hymm reviewed (commented):
Should add some documentation to App::add_systems and Schedule::add_systems that using run_if, after, etc adds a anonymous system set.

cart commented:
@alice-i-cecile:

> I'd like to consolidate this in the future, and make all instances of run conditions with the same name get checked exactly once. Users who want different timing can newtype. I think that the way that each run_if is evaluated seperately in some circumstances is already a footgun in confusing ways.

Hmmm not (yet) sold on that. Having conditions be "assigned" to specific groups of systems, and having them run right before the first system in the group is definitely implicit. But it ties the condition to specific contexts / places in the schedule, which gives them a good chance of being evaluated at the "right time". And the behavior is predictable (and roughly controllable), if you know the rules.

Evaluating them globally removes a lot of "locational context". Definitely a hard problem / one worth discussing. Something we could consider: add "dataflow" to the graph to make these things more clear. Allow scheduling systems that produce outputs (ex: `System<Out = bool>`), which other systems can then depend on. Run conditions would then be configured to run just like any other system. We could have `run_if(some_condition_system)` behave more like `after()` edges. And if we accept arbitrary output types, we could enable more interesting dataflow.

geieredgar commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -555,10 +565,7 @@ impl IntoSystemConfigs<()> for SystemConfigs {
-            config.graph_info.sets.push(set.dyn_clone());
-        }
-
+        self.graph_info.sets.push(Box::new(set));
```
Honestly, I just copy-pasted the `impl IntoSystemConfig<()> for SystemConfig`, but I agree that using `dyn_clone` here is better. I will also fix this for  `SystemConfig`.

cart replied:
Haha I'm responsible for this pattern in this code: https://github.com/bevyengine/bevy/pull/6587#discussion_r1030854349

Forgot about the double-box protection. @maniwani sorry :)

cart replied:
Ah wait no I didn't forget about double-box. We didn't support it yet. Now we do, presumably?

cart replied:
We don't! And we probably shouldn't!

geieredgar replied:
Ah, because `SystemSet` is not implemented for `Box<dyn SystemSet>`, there is no risk of double boxing.
So I can keep the explicit boxes, right?

cart replied:
Correct

cart commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -157,3 +186,39 @@ where
+    use super::AnonymousSet;
+
+    #[test]
+    fn same_instance() {
```
I don't think we need this test anymore, as this is just testing the auto-impled hasher.

geieredgar replied:
Removed.

cart commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -371,8 +375,38 @@ impl ScheduleGraph {
+            conditions,
+            chained,
+        } = systems.into_configs();
+        if conditions.is_empty()
```
I think its worth moving these checks into a `should_add_anonymous_set` method on SystemConfigs, with a comment explaining the purpose. We should also add a comment to the SystemConfigs struct explaining that if new fields are added, they should be considered for inclusion in `should_add_anonymous_set`.

geieredgar replied:
Done.

cart reviewed (commented):
The latest  changes resolve the concerns I listed. I'm on board!

geieredgar commented on crates/bevy_app/src/app.rs:
```diff
@@ -397,9 +397,26 @@ impl App {
+    /// on [`IntoSystemConfigs<P>`] like in the third example above will implicitly add an
+    /// [`AnonymousSet`] that contains all listed systems.
+    ///
+    /// [`AnonymousSet`]: ../ecs/schedule/struct.AnonymousSet.html
```
Im not sure about the `[AnonymousSet]: ../ecs/schedule/struct.AnonymousSet.html` line, but this was the only way I could get my local rustdoc to produce a correct link without having to import the otherwise unused `AnonymousSet` struct. Maybe someone has an better idea?

alice-i-cecile replied:
bevy_ecs::schedule::AnonymousSet should work.

geieredgar replied:
Yeah, I've tried that and rust-analyzer resolves the link correctly, but in the html output of rustdoc an incorrect `<a href="bevy_ecs::schedule::AnonymousSet">` is produced. But maybe this is only an issue when running rustdoc locally. So I changed it to your suggestion, thank you.

geieredgar replied:
Investigating further, the links in docs.rs and the internal dev-docs are also wrong. See the links in https://docs.rs/bevy/0.9.1/bevy/app/struct.App.html#method.register_type as an example.

alice-i-cecile replied:
Oof. That looks like a rustdoc bug? I'd probably leave it as is: people will get the idea and as things get improved these links will all get fixed.

geieredgar replied:
Okay.

geieredgar commented:
I've commited all requested changes (except for the `dyn_clone()` part). Outside my remark about the rustdoc item link I think this is ready for further reviews / merging.

maniwani commented:
`SystemConfigs` was meant to be "distributive", with `add_systems` being equivalent to calling `add_system` multiple times. The `chain` method was an exception (but it only exists on the tuple trait anyway).

I do not like the change from "distributive" to "transitive".

> This is very implicit. There is a behavioral difference between `.add_system(foo.run_if(X)).add_system(bar.run_if(X))` vs `.add_systems((foo, bar).run_if(X))`, which will be more pronounced the more "separate" foo and bar are in a schedule.

Yeah, I think this is behavior too subtle. I'd be onboard if `SystemConfigs` remains "distributive" and either we pick a different method name (i.e. not `run_if`) or add an `is_anonymous_set` (bikeshed) that indicates "special treatment" like `chain` does.

> - It makes run conditions "more optimal" by only running them once for a group of systems.
> - It makes our graphs smaller and therefore faster to "crunch". 

IMO these aren't pros. The act of putting things in a set can give benefits, but we do not want to group systems by accident.

> I'd like to consolidate this in the future, and make all instances of run conditions with the same name get checked exactly once. Users who want different timing can newtype. I think that the way that each run_if is evaluated seperately in some circumstances is already a footgun in confusing ways.

This doesn't really make sense. Take `run_if(in_state(X))` as an example. It should work anywhere in the schedule. IMO the current behavior is clear and intuitive, but I'm obviously biased since I wrote it.

- conditions attached to a system run just before the system does
- conditions attached to a system set run just before any of its systems do

(The "just before" is literal, but the actual invariant is that any data accessed by the conditions is not mutated in between the conditions being evaluated and the (first of the) guarded system(s) running.)

> Run conditions would then be configured to run just like any other system. We could have run_if(some_condition_system) behave more like after() edges. And if we accept arbitrary output types, we could enable more interesting dataflow.

AFAIK dependencies can't (yet) express "before and after + no writer in between". That realization is what led to the current impl. It's the same problem as "before and after + sync point in between". Even if we could do those, having conditions in the graph would lengthen schedule build times (and make visualizations noiser unless they were filtered out).

> Allow scheduling systems that produce outputs (ex: `System<Out = bool>`)

~~systems as entities~~

geieredgar commented:
@maniwani How about adding a `as_set()` method that sets a marker field like `chain()` and documenting that this is currently required for using `run_if`, enforcing it via panic at runtime, if the marker is not set.
This gives users the ability to choose and removes the implicitness.

geieredgar commented:
Clippy was complaining about `as_set` violating naming conventions, so I went with `into_set` instead.

Summarization of my latest changes:

- The behavior is now exactly the same as before the PR, if `into_set()` is not called.
-  A User must explicitly call `into_set()` if he wants to operate on the whole set instead of on each individual system, for example to reduce the amount of edges added to the graph.
- Calling `into_set()` is required if the user wants to use `run_if()`, otherwise a `panic!()` will occur, notifying the user about the misuse of the API.
- The user can instead call `distributive_run_if()` with a `Condition` that implements `Clone`, This will add a clone of the `Condition` to each individual system.
- Inverse to `run_if()`, `distributive_run_if()` must not be called after `into_set()` or an `panic!()` will occur.

alice-i-cecile commented:
Very nice: I like the increase in explicitness. The `distributive_run_if` is a particularly nice touch; it's well-named and captures the subtle difference in behaviors well.

cart commented:
> A User must explicitly call into_set() if they want to operate on the whole set instead of on each individual system, for example to reduce the amount of edges added to the graph.

At this point, is this additional abstraction really "worth it" vs just defining an explicit named set? The change in the behavior of operations if `into_set` is called results in some weird / hard to follow behavior:

```rust
app.add_systems(
  (foo, bar)
     // adds to foo and bar individually
    .after(X)
    .into_set()
    // adds to anonymous set
    .before(Y)
)
```

We could maybe solve this problem by making these ops "lazy"? Ex: store everything on graph_info no matter the state, but then project it onto individual systems if its not a set?

But it also feels like making this opt-in turns this into a niche feature that is "random magic that might make your app perform better". People willing to cast these spells would be equally willing to use explicit sets I think.

```rust
// This is isn't terribly worse than into_set(), and it has the added benefit of:
// 1) Not requiring the introduction of the "anonymous set" concept
// 2) Giving the set a useful name, which can be referenced elsewhere
// 3) Not having implicit behavior changes in ConfigureSystems
#[derive(SystemSet, Clone, PartialEq, Eq Hash, Debug)]
struct MySet;
app
  .configure_set(MySet.after(X).before(Y))
  .add_systems(
    (foo, bar).in_set(MySet)
  )
```

Sorry for the back and forth on this @geieredgar. This is a pretty "central" api and we try to be very careful about such changes.

> into_set

If we do choose to roll with this, maybe we should call this `in_anonymous_set()` or some other `in_X_set()` for parity with `in_set(X)`?

`into_set()` (much like `as_set()`) already means something in the context of Rust / might be a bit confusing.

> distributive_run_if

I do like the explicitness here

geieredgar commented:
@cart and @alice-i-cecile , as both of you liked the explicitness of the `distributive_run_if`, how do you feel about the following proposal:

1. Remove `into_set()` again.
2. Rename `run_if` to `collective_run_if`.
3. Explain in the documentation of `collective_run_if` that the call will result in the creation of an anonymous set that will contain each system and will only run if the passed condition is fulfilled.
4. All other operations will be reverted back, they will behave exactly like before, the anonymous set will only contain the edges to each system and the conditions passed via `collective_run_if`.

alice-i-cecile commented:
I think that's a fairly nice incremental improvement. Still a bit hesitant about anonymous system sets, but I suppose we already have that with type id systems. @jakobhellermann, opinions on this from a schedule viz perspective?

marstaik commented:
Looking forward to this.
I'm new to bevy and by assumption (in regards to how the other commands work ergonomically) I expected `.add_systems((a,b,c).run_if(...)))` to work, and was surprised today to find it didn't already work this way.

Following the discussions I thought I'd drop another idea

```rust
// Regular
pub fn add_system<M>(&mut self, systems: impl IntoSystemAppConfig<M>)

// Applies all .run_if(...), .before(...), .after(...) in a distributive fashion to all systems, so basically clone the conditions to every system
pub fn add_systems<M>(&mut self, systems: impl IntoSystemAppConfigs<M>)

// Create an anonymous set (sys_a, sys_b) such that run_if(...), before(...), .after(...) applies to the anonymous set
pub fn add_system_set<M>(&mut self, systems: impl IntoSystemSetAppConfig<M>)
```

Then I see some weirdness with `.chain()` as being the outlier, which could potentially be replaced with an `ordered()` system set, like so:

```rust
app.add_systems((a,b,c).chain())

// vs

app.add_system_set((a,b,c).ordered())
```

It seems like a good example of future cases where there might be markers such as `ordered` that would be set-specific instead of system specific.

alice-i-cecile commented:
@geieredgar, can you resolve the conflicts or rebase? Once that's done, this should be good to merge.

geieredgar commented:
I've rebased the PR, but I am not sure if it would be better to just change `run_if` instead of adding `collective_run_if`.

alice-i-cecile commented:
> I've rebased the PR, but I am not sure if it would be better to just change `run_if` instead of adding `collective_run_if`.

Agreed, I think this is the way forward (cc @cart @maniwani). We may also want to remove `distributive_run_if`, but that should be completed and discussed seperately IMO.

cart commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -536,9 +544,26 @@ impl ScheduleGraph {
                         }
                     };
+                    let more_than_one_entry = config_iter.len() > 0;
+                    let anonymous_set = if more_than_one_entry && !collective_conditions.is_empty()
```
We never actually add the collective conditions anywhere so run_if doesn't work at all here.

cart replied:
Fixed in my commit below

cart commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -536,9 +544,26 @@ impl ScheduleGraph {
+                    } else {
+                        None
+                    };
+                    let prev = if let Some(anonymous_set) = anonymous_set {
```
This repeated code can be avoided if we operate directly on the configs Vec instead of doing it in the chain code.

cart replied:
Fixed in my commit below

cart commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -536,9 +544,26 @@ impl ScheduleGraph {
                         }
                     };
+                    let more_than_one_entry = config_iter.len() > 0;
+                    let anonymous_set = if more_than_one_entry && !collective_conditions.is_empty()
```
This doesn't handle the case where we have collective conditions in a single SystemConfigs::Configs (ex: `(foo, ).run_if(X)`

cart replied:
Fixed in my commit below

cart reviewed (commented):
I think this is the right general approach, although the implementation doesn't actually work atm (see comments).

I have fixes prepared that I'll push now.
